File: 34-Search-for-a-Range/solution.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ret = [-1, -1]
        if not nums:
            return ret

        left, right = 0, len(nums) - 1

        while left + 1 < right:
            mid = (left + right) / 2
            if nums[mid] == target:
                right = mid
            elif nums[mid] < target:
                left = mid
            else:
                right = mid

        if nums[left] == target:
            ret[0] = left
        elif nums[right] == target:
            ret[0] = right

        left, right = 0, len(nums) - 1

        while left + 1 < right:
            mid = (left + right) / 2

            if nums[mid] == target:
                left = mid
            elif nums[mid] < target:
                left = mid
            else:
                right = mid

        if nums[right] == target:
            ret[1] = right
        elif nums[left] == target:
            ret[1] = left

        return ret
        

File: 34-Search-for-a-Range/solution.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        const int begin = lower_bound(nums, target);
        const int end = upper_bound(nums, target);

        if (begin < nums.size() && nums[begin] == target)
            return {begin, end - 1};

        return {-1, -1};
    }
    
private:
    int lower_bound(vector<int> &nums, int target) {
        int left = 0;
        int right = nums.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] >= target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    int upper_bound(vector<int> &nums, int target) {
        int left = 0;
        int right = nums.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
};